landoapi/hg.py

```diff
@@ -321,6 +336,9 @@ def apply_patch(self, patch_io_buf):
             )
 
     def push(self, target, bookmark=None):
+        if not os.getenv(REQUEST_USER_ENV_VAR):
+            raise ValueError(f"{REQUEST_USER_ENV_VAR} not set while attempting to push")
```

Comment:
This state-check is giving a little smell that perhaps the handle for pushing/pulling should be contextual. This ensures that functions like `push()` here can only be called when the appropriate handle has been instantiated, e.g.:
```
class HgPushHandle():
    def __init__(repo):
        self._repo = repo
    
    def push(target, bookmark=None):
        # perhaps set the env here?


class HgRepo():
    # ...
    @contextmanager
    def for_push():
        hg_handle = self._open()
        hg_push_handle = HgPushHandle(hg_handle)
        yield hg_push_handle
        self._clean_and_close(hg_handle)
```

Of course, there's a balance here - increased weight on a Python structural system, but better guard rails against API misuse. I like to lean more on better guard rails, but I'm completely on board if you feel that it adds too many types and that it may not be worth it :)

Reply:
I think this is out of scope for this PR, but is a good suggestion, though it might need some tweaking. The `for_pull` context manager doesn't only provide the push functionality, it also will provide everything else the `for_pull` context manager does. It would be kind of awkward to do all of that via the "push handle".

Reply:
True, though if the `for_push` handle is only used for pushing, it's ok to strip it of non-push functionality, even though it's technically possible. I think that you have better context here, so :+1: from me!